src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java

```diff
@@ -209,34 +225,114 @@ private void loadConfig(Map<String, Object> config, AmazonCloudWatch client) {
+    private List<ResourceTagMapping> getResourceTagMappings(MetricRule rule, AWSResourceGroupsTaggingAPI taggingClient) {
+	List<ResourceTagMapping> resourceTagMappings = new ArrayList<ResourceTagMapping>();
+	if(rule.awsTagSelect == null) {
+            return resourceTagMappings;
+        }
+	
+        List<TagFilter> tagFilters = new ArrayList<TagFilter>();
+        for (Map.Entry<String, List<String>> entry : rule.awsTagSelect.tagSelections.entrySet()) {
+          tagFilters.add(new TagFilter().withKey(entry.getKey()).withValues(entry.getValue()));
+        }
+
+        GetResourcesRequest request = new GetResourcesRequest().withTagFilters(tagFilters).withResourceTypeFilters(rule.awsTagSelect.resourceTypeSelection);
+        String paginationToken = "";
+        do {
+          request.setPaginationToken(paginationToken);
+          
+          GetResourcesResult result = taggingClient.getResources(request);
+          resourceTagMappings.addAll(result.getResourceTagMappingList());
+          
+          paginationToken = result.getPaginationToken();
+        } while (paginationToken != null && paginationToken != "");
+     
+        return resourceTagMappings;
+    }
+    
+    private List<String> extractResourceIds(List<ResourceTagMapping> resourceTagMappings) {
+	List<String> resourceIds = new ArrayList<String>();
+	for (ResourceTagMapping resourceTagMapping : resourceTagMappings) {
+	    resourceIds.add(extractResourceId(resourceTagMapping.getResourceARN()));
+	}
+	return resourceIds;
+    }
+    
+    private String extractResourceId(String arn) {
+	// ARN parsing is based on https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+	String[] arnArray = arn.split(":");
+	String resourceId = arnArray[arnArray.length-1];
+	if(resourceId.contains("/")) {
+	    String[] resourceArray = resourceId.split("/", 2);
+	    resourceId = resourceArray[resourceArray.length-1];
+	}
+
+	return resourceId;
+    }
+    
+    private List<List<Dimension>> getDimensions(MetricRule rule, List<String> tagBasedResourceIds, AmazonCloudWatch cloudWatchClient) {
         if (
                 rule.awsDimensions != null &&
                 rule.awsDimensionSelect != null &&
                 rule.awsDimensions.size() > 0 &&
                 rule.awsDimensions.size() == rule.awsDimensionSelect.size() &&
-                rule.awsDimensionSelect.keySet().containsAll(rule.awsDimensions)
+                rule.awsDimensionSelect.keySet().containsAll(rule.awsDimensions) &&
+                rule.awsTagSelect == null
         ) {
             // The full list of dimensions is known so no need to request it from cloudwatch.
             return permuteDimensions(rule.awsDimensions, rule.awsDimensionSelect);
         } else {
-            return listDimensions(rule, client);
+            return listDimensions(rule, tagBasedResourceIds, cloudWatchClient);
         }
```

Comment:
I know it wasn't introduced by your change, but I would remove the else statement if possible

Reply:
@rolandvarga  Except if it is very very important for you, I would not change that to avoid confusion.